catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/QueryResultPaginator.java

```diff
@@ -0,0 +1,160 @@
+import ddf.catalog.source.SourceUnavailableException;
+import ddf.catalog.source.UnsupportedQueryException;
+
+/**
+ * This class controls paging of {@link CatalogFramework} queries by batching them into lists of
+ * results, independent of the number of results the Catalog Framework returns with each query.
+ * This helps clients guarantee a particular batch size rather than getting fewer results than
+ * expected from the Catalog Framework.
+ */
+public class QueryResultPaginator {
+
+    private final CatalogFramework catalogFramework;
+
+    // Used to query the catalog framework and adjust the starting index of the query without
+    // modifying the original query.
+    private final QueryImpl queryCopy;
+
+    // Used to track the start index of the next query and ensure that the correct query results
+    // are returned with no repetition of results.
+    private int indexOffset;
+
+    private int pageSize;
+
+    // Set to true when a query returns zero results and no more results can be fetched from
+    // Catalog Framework..
+    private boolean noMoreCatalogFrameworkResultsToFetch;
+
+    // Used to buffer fetched results from the Catalog Framework and hold them until they can be
+    // drained into a result list to return to the client.
+    private BlockingQueue<Result> queriedResultsBuffer;
+
+    /**
+     * Creates a paginator based on the {@link Query} provided.
+     * <p>
+     * The page size provided in the {@link Query} is the guaranteed size of any list of results
+     * returned to a client.
+     * </p>
+     *
+     * @param catalogFramework reference to the {@link CatalogFramework}
+     * @param query            client query. Will be used to query the {@link CatalogFramework}
+     *                         and guarantee that the number of results returned will always match
+     *                         the query page size provided until all results have been retrieved.
+     */
+    public QueryResultPaginator(CatalogFramework catalogFramework, Query query) {
+        Validate.notNull(catalogFramework);
+        Validate.notNull(query);
+
+        this.catalogFramework = catalogFramework;
+        this.queryCopy = copy(query);
+
+        noMoreCatalogFrameworkResultsToFetch = false;
+        pageSize = queryCopy.getPageSize();
+        indexOffset = queryCopy.getStartIndex();
+        queriedResultsBuffer = new LinkedBlockingDeque<>();
+    }
+
+    public boolean hasNext() {
+
+        if (notEmpty()) {
+            return true;
+        }
+
+        if (noMoreCatalogFrameworkResultsToFetch) {
+            return false;
+        }
+
+        // Buffer is empty, but query might not be exhausted.
+        fetchNext();
+
+        return notEmpty();
+    }
+
+    public List<Result> next() throws NoSuchElementException {
+        if (!hasNext()) {
+            throw new NoSuchElementException();
+        }
+
+        while (!noMoreCatalogFrameworkResultsToFetch && getBufferSize() < pageSize) {
```

Comment:
It's unfortunate that we have to handle both the "more than pageSize returned" and "less than pageSize returned" scenarios.